File: src/belnorm/casing.py
"""Capitalisation patterns: detect one on the input word, re-apply it to the output.

Conversion always works on lowercase forms (the lexicon is stored lowercase
and rules are written lowercase), so the original case has to survive
round-tripping through the cascade.
"""

from __future__ import annotations

from typing import Literal

CaseForm = Literal["lower", "upper", "title", "mixed"]


def detect_case(word: str) -> CaseForm:
    letters = [c for c in word if c.isalpha()]
    if not letters:
        return "lower"
    if all(c.islower() for c in letters):
        return "lower"
    if letters[0].isupper() and all(c.islower() for c in letters[1:]):
        return "title"
    if all(c.isupper() for c in letters):
        # A single capital letter ("Я", "У") is ambiguous; treat it as title
        # so "У Менску" does not come back as "У МЕНСКУ".
        return "title" if len(letters) == 1 else "upper"
    return "mixed"


def apply_case(word: str, form: CaseForm) -> str:
    match form:
        case "lower":
            return word
        case "upper":
            return word.upper()
        case "title":
            for i, c in enumerate(word):
                if c.isalpha():
                    return word[:i] + c.upper() + word[i + 1 :]
            return word
        case "mixed":
            return word


def recase(source: str, target: str) -> str:
    """Give ``target`` (lowercase) the capitalisation pattern of ``source``."""
    return apply_case(target, detect_case(source))

File: js/src/casing.ts
// Capitalisation patterns: detect one on the input word, re-apply it to the output.
// Mirrors src/pravapis/casing.py.

export type CaseForm = "lower" | "upper" | "title" | "mixed";

const isAlpha = (c: string): boolean => /\p{L}/u.test(c);
const isLower = (c: string): boolean => c === c.toLowerCase() && c !== c.toUpperCase();
const isUpper = (c: string): boolean => c === c.toUpperCase() && c !== c.toLowerCase();

export function detectCase(word: string): CaseForm {
  const letters = [...word].filter(isAlpha);
  if (letters.length === 0) return "lower";
  if (letters.every(isLower)) return "lower";
  if (isUpper(letters[0]!) && letters.slice(1).every(isLower)) return "title";
  if (letters.every(isUpper)) {
    // A single capital letter ("Я", "У") is ambiguous; treat it as title so
    // "У Менску" does not come back as "У МЕНСКУ".
    return letters.length === 1 ? "title" : "upper";
  }
  return "mixed";
}

export function applyCase(word: string, form: CaseForm): string {
  switch (form) {
    case "lower":
      return word;
    case "upper":
      return word.toUpperCase();
    case "title": {
      for (let i = 0; i < word.length; i++) {
        if (isAlpha(word[i]!)) {
          return word.slice(0, i) + word[i]!.toUpperCase() + word.slice(i + 1);
        }
      }
      return word;
    }
    case "mixed":
      return word;
  }
}

/** Give `target` (lowercase) the capitalisation pattern of `source`. */
export function recase(source: string, target: string): string {
  return applyCase(target, detectCase(source));
}
